File: backend/app/models/schemas.py
from datetime import datetime
from enum import Enum
from typing import Literal

from pydantic import BaseModel, Field, HttpUrl


class JobStatus(str, Enum):
    queued = "queued"
    running = "running"
    done = "done"
    failed = "failed"
    timeout = "timeout"
    invalid_repo = "invalid_repo"


class ProgressStage(str, Enum):
    ingest = "ingest"
    classify = "classify"
    detect = "detect"
    graph = "graph"
    narrate = "narrate"


class RoleTag(str, Enum):
    initiator = "Initiator"
    implementer = "Implementer"
    corrector = "Corrector"
    unblocker = "Unblocker"


class DecisionType(str, Enum):
    convergence = "convergence"
    replacement = "replacement"
    unblocking = "unblocking"
    correction = "correction"
    architectural_pivot = "architectural_pivot"


class RelationType(str, Enum):
    builds_on = "builds_on"
    unblocks = "unblocks"
    corrects = "corrects"


class ErrorCode(str, Enum):
    invalid_repo = "invalid_repo"
    clone_failed = "clone_failed"
    parse_failed = "parse_failed"
    timeout = "timeout"
    llm_schema_error = "llm_schema_error"
    internal_error = "internal_error"
    not_found = "not_found"
    not_ready = "not_ready"


class AnalysisCreateRequest(BaseModel):
    repo_url: HttpUrl


class AnalysisCreateResponse(BaseModel):
    job_id: str
    status: JobStatus
    created_at: datetime


class AnalysisJobResponse(BaseModel):
    job_id: str
    status: JobStatus
    created_at: datetime
    updated_at: datetime
    progress_stage: ProgressStage | None = None
    message: str | None = None


class ErrorDetail(BaseModel):
    code: ErrorCode | str
    message: str
    details: str | None = None
    retryable: bool = False


class ErrorEnvelope(BaseModel):
    error: ErrorDetail


class EvidenceRef(BaseModel):
    commit_sha: str = Field(min_length=7, max_length=64)
    file_paths: list[str] = Field(default_factory=list)
    reason: str = Field(min_length=3)


class DecisionMoment(BaseModel):
    decision_id: str
    type: DecisionType
    title: str
    summary: str
    contributor_id: str
    timestamp: datetime
    confidence: float = Field(ge=0.0, le=1.0)
    evidence: list[EvidenceRef] = Field(min_length=1)


class GraphNode(BaseModel):
    id: str
    kind: Literal["contributor", "decision"]
    label: str
    influence_score: float = Field(ge=0.0)


class GraphEdge(BaseModel):
    id: str
    source: str
    target: str
    weight: float = Field(ge=0.0)
    relation: RelationType


class GraphPayload(BaseModel):
    nodes: list[GraphNode]
    edges: list[GraphEdge]


class ContributorEvidence(BaseModel):
    commit_sha: str
    reason: str


class ContributorProfile(BaseModel):
    contributor_id: str
    name: str
    role: RoleTag
    narrative: str
    confidence: float = Field(ge=0.0, le=1.0)
    top_decisions: list[str]
    evidence: list[ContributorEvidence] = Field(min_length=1)


class ProjectSummary(BaseModel):
    name: str
    repo_url: str
    commit_count: int
    contributors_count: int


class AnalysisResultResponse(BaseModel):
    job_id: str
    project: ProjectSummary
    decisions: list[DecisionMoment]
    graph: GraphPayload
    contributors: list[ContributorProfile]
    warnings: list[str]

File: lib/schemas.ts
import { z } from "zod";

export const jobStatusSchema = z.enum([
  "queued",
  "running",
  "done",
  "failed",
  "timeout",
  "invalid_repo",
]);

export const progressStageSchema = z.enum(["ingest", "classify", "detect", "graph", "narrate"]);

export const analysisJobSchema = z.object({
  job_id: z.string(),
  status: jobStatusSchema,
  created_at: z.string(),
  updated_at: z.string(),
  progress_stage: progressStageSchema.nullish().transform((v) => v ?? undefined),
  message: z.string().nullish().transform((v) => v ?? undefined),
  error_code: z.string().nullish().transform((v) => v ?? undefined),
  error_message: z.string().nullish().transform((v) => v ?? undefined),
});

export const analysisCreateResponseSchema = z.object({
  job_id: z.string(),
  status: jobStatusSchema,
  created_at: z.string(),
});

const evidenceSchema = z.object({
  commit_sha: z.string(),
  file_paths: z.array(z.string()).optional(),
  reason: z.string(),
});

const decisionSchema = z.object({
  decision_id: z.string(),
  type: z.enum(["convergence", "replacement", "unblocking", "correction", "architectural_pivot"]),
  title: z.string(),
  summary: z.string(),
  contributor_id: z.string(),
  timestamp: z.string(),
  confidence: z.number(),
  evidence: z.array(evidenceSchema),
});

const graphNodeSchema = z.object({
  id: z.string(),
  kind: z.enum(["contributor", "decision"]),
  label: z.string(),
  influence_score: z.number(),
});

const graphEdgeSchema = z.object({
  id: z.string(),
  source: z.string(),
  target: z.string(),
  weight: z.number(),
  relation: z.enum(["builds_on", "unblocks", "corrects"]),
});

const contributorSchema = z.object({
  contributor_id: z.string(),
  name: z.string(),
  role: z.enum(["Initiator", "Implementer", "Corrector", "Unblocker"]),
  narrative: z.string(),
  confidence: z.number(),
  top_decisions: z.array(z.string()),
  evidence: z.array(
    z.object({
      commit_sha: z.string(),
      reason: z.string(),
    }),
  ),
});

export const analysisResultSchema = z.object({
  job_id: z.string(),
  project: z.object({
    name: z.string(),
    repo_url: z.string(),
    commit_count: z.number(),
    contributors_count: z.number(),
  }),
  decisions: z.array(decisionSchema),
  graph: z.object({
    nodes: z.array(graphNodeSchema),
    edges: z.array(graphEdgeSchema),
  }),
  contributors: z.array(contributorSchema),
  warnings: z.array(z.string()),
});

export const analysisSummarySchema = z.object({
  job_id: z.string(),
  project: z.object({
    name: z.string(),
    repo_url: z.string(),
    commit_count: z.number(),
    contributors_count: z.number(),
  }),
  warnings: z.array(z.string()),
});

export const timelineSchema = z.object({
  job_id: z.string(),
  decisions: z.array(decisionSchema),
});

export const contributorResponseSchema = z.object({
  job_id: z.string(),
  contributor: contributorSchema,
});

export const contributorsResponseSchema = z.object({
  job_id: z.string(),
  contributors: z.array(contributorSchema),
});
